docs/dependency-management.md

```diff
@@ -0,0 +1,50 @@
+# Dependency Management Guidelines
+
+This repo uses a number of third-party rust crates, so we need to be mindful
+of how these dependencies will affect our consumers.  Considerations include:
+
+* General code quality
+* [Licensing compatibility](https://www.mozilla.org/en-US/MPL/license-policy/#Licenses_Compatible_with_the_MPL)
+* Handling of security vulnerabilities
+* The potential for [supply-chain compromise](https://medium.com/intrinsic/compromised-npm-package-event-stream-d47d08605502)
+
+We're still evolving our policies in this area, but here are the
+guidelines we've developed so far:
+
+* When adding a new dependency:
+  * In general, be conservative in adding new third-party dependencies.
+    * For trivial functionality, consider just writing it yourself.
+      Remember the cautionary tale of [left-pad](https://www.theregister.co.uk/2016/03/23/npm_left_pad_chaos/).
+    * Check if we already have a crate in our dependency tree that can provide the needed functionality.
+    * Prefer widely-used crates or crates from rust-lang-nursery.
+  * Check that it is clearly licensed and is [MPL-2.0 compatible](https://www.mozilla.org/en-US/MPL/license-policy/#Licenses_Compatible_with_the_MPL).
+  * Take the time to investigate the crate's source and ensure it is suitably high-quality.
+    * Be especially wary of uses of `unsafe`.
+    * Use `cargo fetch` to obtain the same source code that cargo will eventually build.
+  * Transitively apply this reasoning to all of its dependencies.
+  * Always depend on a specific version of the dependency, avoiding semver pattern-matching.
+  * Open a separate PR just for introducing the dependency.
+    * (Editor's note: I feel like this may be controversial but wanted to put it out there for consideration...)
+    * This will help provide a clear audit trail for when and why dependencies were introduced.
+    * It will also help us remember to take these steps seriously.
+  * Dev dependencies do not require as much scrutiny as dependencies that will ship in consuming applications,
```

Comment:
The difficulty here is mainly that cargo doesn't provide a lot of insight to what deps are build/dev-deps. IIUC it considers it a single dependency tree under the hood (same for target-specific deps).

Reply:
Hmm... A lot of our weirder dev-dependencies are from use in the samples. I wonder if it would be worth something like: moving them to a `samples` dir which is `exclude`ed from our workspace's Cargo.toml.

The downside here is that it makes 'check for compilation errors' harder. Now we need to run cargo commands from both the samples and the local dir.

We could address this maybe by having a phase for `cargo check --all --all-targets --all-features` in both workspaces and having it be required to pass for landing (our extended circleci bustage has made us pretty willing to ignore ci failures so I do think this might be required). But that's still inconvenient. I'm beginning to see why complex projects often end up with their own `./mach`-like tools.

That said, I think the only thing that's different in terms of use with dev-dependencies are our threshold for living without it. For example, I wouldn't use pretty-assertations in a release dep, but I would in a dev dep.